### PR TITLE
[BEAM-894] Using @Teardown to remove temp files from failed bundles in Write.WriteBundles.

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi @tgroh

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java [hunk 1]
```diff
@@ -259,6 +259,14 @@ public abstract void finalize(Iterable<WriteT> writerResults, PipelineOptions op
     public abstract WriteT close() throws Exception;
 
     /**
+     * Aborts writing the bundle. Removes resources used for writing the bundle,
+     * and closes this {@link Writer}.
+     *
+     * <p>Called for each failed bundle.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: I do not think this is a correct description of when this function will be called, based on the semantics of TearDown.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java [hunk 2]
```diff
@@ -207,6 +191,14 @@ public void finishBundle(Context c) throws Exception {
         }
       }
 
+      @Teardown
+      public void teardown() throws Exception {
+        if (writer != null) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]: add a comment about why this is correct.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java [hunk 3]
```diff
@@ -103,4 +103,9 @@
 
   /** Converts the given string to a {@link Path}. */
   Path toPath(String path);
+
+  /**
+   * Delete files with the given {@code uris} if they exist.
+   */
+  void deleteIfExists(Collection<String> uris) throws IOException;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java]: do you really want to add this API? Why not move this logic to the caller of delete?

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java [hunk 4]
```diff
@@ -265,6 +265,11 @@ public String close() throws Exception {
     }
 
     @Override
+    public void abort() throws Exception {
+      close();
```
Comment (dhalperi) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java]: why is this one not deleting the file?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java [hunk 5]
```diff
@@ -569,6 +569,11 @@ public TestWriterResult close() throws Exception {
       state = State.CLOSED;
       return new TestWriterResult(uId, elementsWritten);
     }
+
+    @Override
+    public void abort() throws Exception {
+      close();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: add assertions about state. actually test this.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 6]
```diff
@@ -641,6 +641,12 @@ public final FileResult close() throws Exception {
       return result;
     }
 
+    @Override
+    public final void abort() throws Exception {
+      IOChannelUtils.deleteIfExists(filename);
+      close();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: For some file systems (like GCS), close is what actually creates the file.

you should close first, then delete with failure handling after close finishes.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java [hunk 7]
```diff
@@ -178,23 +178,7 @@ public void processElement(ProcessContext c) throws Exception {
...
-            if (closeException instanceof InterruptedException) {
-              // Do not silently ignore interrupted state.
-              Thread.currentThread().interrupt();
-            }
-            // Do not mask the exception that caused the write to fail.
-            e.addSuppressed(closeException);
-          }
-          throw e;
-        }
+        writer.write(c.element());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]: this is a nice improvement.

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java (hunk 4 above)
Comment (peihe) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java]: I don't familiar with this code, and looks to me there is no tests. So, I prefer not to change it in this PR.

(I made the change deleting the file here, but stashed it.)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: We do not guarantee that the `@Teardown` method will be called on a failed bundle (e.g. what happens if the worker crashes?) The only strong guarantee is that no `DoFn` method will be called on an instance after the call to the `@Teardown` method. We will -usually- call `@Teardown` if a bundle fails, but cannot make guarantees to that effect.

Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: Update javadoc to clarify that it is best effort.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java [hunk 8]
```diff
@@ -259,6 +259,15 @@ public abstract void finalize(Iterable<WriteT> writerResults, PipelineOptions op
     public abstract WriteT close() throws Exception;
 
     /**
+     * Aborts writing the bundle. Removes resources used for writing the bundle,
+     * and closes this {@link Writer}.
+     *
+     * <p>Called for each failed bundle with best effort. However, there is no strong guarantee in
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: I would reword this to something more like

"When a bundle fails, the runner will `abort` the Writer if possible."
and link to `@Teardown` as a reference.

I don't think we need to give examples of when that will occur - just that we'll try but can't always succeed. It may be worth also noting that additional cleanup can be done as part of the `finalize` step in case `abort` could not be called on all failed bundles.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java]: Generally we should provide a tightly scoped primitive operation in `IOChannelFactory` and feel free to provide utilities around it - so this can be `delete` and throw a `FileNotFoundException` if the underlying IO cannot delete it, and then we can reuse the "try delete and suppress `FileNotFoundExceptions`

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java]: Is there a JIRA for adding tests to `HDFSFileSink`?

Comment (peihe) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java]: filed https://issues.apache.org/jira/browse/BEAM-949

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java (hunk 3 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java]: We are doing bulk deletes here. deleteAndThrowIfMissing() is hard to use as a primitive operation. That would requires returning 1. missing file names and 2. files deleted successfully.

This is why I picked deleteIfExists() as the primitive.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java (hunk 8 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: updated

**Comment by peihe** [comment]
PTAL

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java (hunk 3 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java]: I think the actual problem is that you modified `IOChannelFactory` rather than `FileOperations` in `FileBasedSink`, which is the place this code actually belong. You are right that `IOChannelFactory` does not have any delete functionality.

That already has a `remove` function you can use. Perhaps as-is, perhaps with a try and catch.

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelFactory.java]: (Of course in the new code you're working on, `FileOperations` and `IOChannelFactory` will merge into the same class.)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 9]
```diff
@@ -641,6 +641,12 @@ public final FileResult close() throws Exception {
       return result;
     }
 
+    @Override
+    public final void abort() throws Exception {
+      close();
+      IOChannelUtils.deleteIfExists(filename);
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java [hunk 10]
```diff
@@ -311,14 +346,35 @@ public String apply(Long input) {
...
+      PTransform<PCollection<String>, PCollection<String>> transform,
+      boolean failInWrite) {
+    runShardedWriteInternal(inputs, transform, Optional.<Integer>absent(), failInWrite);
+  }
+
+  private static void runShardedWrite(
+      List<String> inputs,
+      PTransform<PCollection<String>, PCollection<String>> transform,
+      Optional<Integer> numConfiguredShards) {
+     runShardedWriteInternal(inputs, transform, numConfiguredShards, false /* failInWRite */);
```
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: done

**Comment by peihe** [comment]
The leftover temp files issue is mitigated by https://github.com/apache/incubator-beam/pull/1278 now.

I am closing this PR since the benefits doesn't justify the cost for the additional method that Sink writers have to implement.
